zkevm-circuits/src/super_circuit.rs

```diff
@@ -334,28 +553,49 @@ impl<F: Field, const MAX_TXS: usize, const MAX_CALLDATA: usize, const MOCK_RANDO
         challenges: &Challenges<Value<F>>,
         layouter: &mut impl Layouter<F>,
     ) -> Result<(), Error> {
-        self.keccak_circuit
-            .synthesize_sub(&config.keccak_circuit, challenges, layouter)?;
-        self.bytecode_circuit
-            .synthesize_sub(&config.bytecode_circuit, challenges, layouter)?;
-        self.tx_circuit
-            .synthesize_sub(&config.tx_circuit, challenges, layouter)?;
-        self.state_circuit
-            .synthesize_sub(&config.state_circuit, challenges, layouter)?;
-        self.copy_circuit
-            .synthesize_sub(&config.copy_circuit, challenges, layouter)?;
-        self.exp_circuit
-            .synthesize_sub(&config.exp_circuit, challenges, layouter)?;
-        self.evm_circuit
-            .synthesize_sub(&config.evm_circuit, challenges, layouter)?;
-        self.pi_circuit
-            .synthesize_sub(&config.pi_circuit, challenges, layouter)?;
+        if let Some(circuit_config) = &config.keccak_circuit {
```

Comment:
I think this needs to handle more conditions.
We have a list of circuits and a list of tables.
Some circuits have dependencies on tables, and you encoded that in `SuperCircuitConfig::new` with flag checks for each table construction, to make sure the tables exist whenever some circuit that uses them is enabled.
Then whenever a table is required by a circuit, the table should be assigned.  If the circuit associated with the table is enabled, then the circuit itself has the assignment code which is in `sub_circuit.synthesize_sub`.  But when that circuit is not enabled but yet the table is needed, we still need to assign it.  For that we have methods to assign the table outside of it's corresponding circuit.  Here's an example of how the Bytecode test circuit assigns the keccak table:
https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/3aa36a2ca9360723a5eb87816a6b02655598d2d5/zkevm-circuits/src/bytecode_circuit/dev.rs#L49-L53

So I think each `if` here requires an else.
For example in the keccack circuit the else should do:
```rust
} else if is_enabled(
    flags,
    SUPER_CIRCUIT_FLAG_TX
    | SUPER_CIRCUIT_FLAG_BYTECODE
    | SUPER_CIRCUIT_FLAG_EVM,
    ) {
        config.keccak_table.dev_load(
            &mut layouter,
            self.bytecodes.iter().map(|b| &b.bytes),
            &challenges,
        )?;
}
```